pkg/session/bootstrap.go

```diff
@@ -3216,6 +3217,13 @@ func doDDLWorks(s sessiontypes.Session) {
 // doBootstrapSQLFile executes SQL commands in a file as the last stage of bootstrap.
 // It is useful for setting the initial value of GLOBAL variables.
 func doBootstrapSQLFile(s sessiontypes.Session) error {
+	if s.GetExtensions() == nil {
```

Comment:
This fix seems a little tricky. Before this PR, all internal session seems do not have a `SessionExtension` set. This PR only sets the session's inner extension when we use it to bootstrap the SQL file that makes some differences of sessions in the pool. Though it will not cause any bug now, but in the future if the extension in internal session has some bug, it will be hard to locate and debug.

A better way is to make sure all internal SQLs are created with internal extension, but seems it has many code to change and to test.

How about using a global extension by `extension.GetExtensions()` when create/alter a user?

Reply:
> How about using a global extension by extension.GetExtensions() when create/alter a user?

My concern for this is `extension.GetExtensions()` requires locking every time and may cause performance issue when create/alter user is getting called with high concurrency.

Maybe instead of using the shared session `ses[9]` for bootstrapping, we can create a new session inside https://github.com/pingcap/tidb/blob/c4fd1b212c0ca6572e3ad6f18fdcdd58354f629c/pkg/session/session.go#L3559
with extensions bind to it specifically for bootstrap usage?

Reply:
Performance isn't a big deal, IMO. You can see that it uses a read-write lock:

https://github.com/pingcap/tidb/blob/2da324d1a1831c7be31ef17a164ff511757cdbb5/pkg/extension/registry.go#L45-L57

For most times, the calls of `GetExtensions` will not block each other because the inner state has been setup and the only read lock is used.

Compared with other parts of DDL such as the network, the cost of `GetExtensions` is insignificant.